src/NUnitEngine/nunit.engine/Services/ExtensionService.cs

```diff
@@ -0,0 +1,154 @@
+﻿// ***********************************************************************
+// Copyright (c) 2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Xml;
+using Mono.Cecil;
+using NUnit.Engine.Extensibility;
+using NUnit.Engine.Internal;
+
+namespace NUnit.Engine.Services
+{
+    public class ExtensionService : Service
+    {
+        private List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
+        private Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
+        private Dictionary<Type, ExtensionPoint> _typeIndex = new Dictionary<Type, ExtensionPoint>();
+
+        private List<ExtensionNode> _extensions = new List<ExtensionNode>();
+
+        #region Public Properties
+
+        public IEnumerable<ExtensionPoint> ExtensionPoints
+        {
+            get { return _extensionPoints; }
+        }
+
+        public IEnumerable<ExtensionNode> Extensions
+        {
+            get { return _extensions;  }
+        }
+
+        #endregion
+
+        public ExtensionPoint GetExtensionPoint(string path)
```

Comment:
Everywhere we get extensions, we do it based on the interface, not the path, so this seems unused and unneeded. By extension, I don't think the path is needed in the ExtensionAttribute. If anything, the attribute should specify the interface that an extension is providing an extension for. Hard coded string paths seem cryptic and error prone to me.